FHEM/00_SIGNALduino.pm

```diff
@@ -2187,95 +2187,112 @@ sub SIGNALduino_Parse_MU($$$$@)
 			my $bit_length = ($signal_length/((scalar @{$ProtocolListSIGNALduino{$id}{one}} + scalar @{$ProtocolListSIGNALduino{$id}{zero}})/2));
 			Debug "Expect $bit_length bits in message"  if ($valid && $debug);
 
-			#Check calculated min length
-			#$valid = $valid && $ProtocolListSIGNALduino{$id}{length_min} <= $bit_length if (exists $ProtocolListSIGNALduino{$id}{length_min}); 
-			#Check calculated max length
-			#$valid = $valid && $ProtocolListSIGNALduino{$id}{length_max} >= $bit_length if (exists $ProtocolListSIGNALduino{$id}{length_max});
-
-			#next if (!$valid) ;
-			#Debug "expecting $bit_length bits in signal" if ($debug && $valid);
-
 			Debug "Searching in patternList: ".Dumper(\%patternList) if($debug);
 			next if (!$valid) ;
 
-	
+			my @msgStartLst;
+			my $startStr="";
+			my $message_start=0 ;
+			my $startLogStr="";
+			
+			if (defined($ProtocolListSIGNALduino{$id}{start}))	# wenn start definiert ist, dann startStr ermitteln und in rawData suchen und in der rawData alles bis zum startStr abschneiden
+			{
+				@msgStartLst = $ProtocolListSIGNALduino{$id}{start};
+				Debug "msgStartLst: ".Dumper(@msgStartLst)  if ($debug);
+				
+				if ( ($startStr=SIGNALduino_PatternExists($hash,@msgStartLst,\%patternList,\$rawData)) eq -1)
+				{
+					SIGNALduino_Log3 $name, 5, "$name: start pattern for MU Protocol id $id -> $ProtocolListSIGNALduino{$id}{name} mismatches, aborting"  ;
+					$valid=0;
+					next;
+				};
+				Debug "startStr is: $startStr" if ($debug);
+				
+				if ($rawData =~ m/$startStr/) {
```

Comment:
wäre der code leichter zu verstehen wenn wir das suchen und ausschneiden durch folgende Regex ersetzen? Das schneidet alles ab, was vor $startStr ist.

($rawData) = $rawData=~ /($startStr.*)$

Reply:
@Ralf9 
noch eine. Hoffe Du weisst jetzt wo die Fragen zu finden sind :)

Reply:
Ich finde dies
`($rawData) = $rawData=~ /($startStr.*)$`
komplizierer und schwieriger zu verstehen als meine Variante.
Was steht in $rawData, wenn es keinen match gibt?
Wie frage ich ab ob es einen match gegeben hat?

Reply:
Jetzt ist mir klar wie das mit
`($rawData) = $rawData=~ /($startStr.*)$`
funktioniert.
Ich sehe da aber keine Vorteile zu meiner Variante
```
				if ($rawData =~ m/$startStr/) {
					$message_start = $+[0];
					$rawData = substr($rawData, $message_start);
					$startLogStr = "Start: $startStr cut Pos $message_start" . "; ";
					Debug "rawData = $rawData" if ($debug);
					Debug "startStr $startStr found. Message starts at $message_start" if ($debug);
				} else {
					Debug "startStr $startStr not found." if ($debug);
					next;
				}
```

Mit Deiner Variante wird es nicht einfacher, es sind genausoviel Zeilen notwendig.
```
($rawData) = $rawData=~ /($startStr.*)$
if ($rawData) { 		# $startStr 
	$message_start = $+[0];
	$startLogStr = "Start: $startStr cut Pos $message_start" . "; ";
	Debug "rawData = $rawData" if ($debug);
	Debug "startStr $startStr found. Message starts at $message_start" if ($debug);
} else {
	Debug "startStr $startStr not found." if ($debug);
	next;
}
```

Bei einer Startseqeuenz von 10 und
2030201020203030202030
ist bei Dir 
1020203030202030
Es steht noch die Startseqeuenz von 10 am Anfang, durch die 10 am Anfang wird die Weiterarbeit aufwendiger.

Reply:
Der Vorteil das Ergebnis aus einer Regex mit einer capture Group liegt einfach darin, das Ergebnis zu verwenden.
Ich bin auch aktuell verwirrt, wann welche Regex geprüft wird.
Ob es startstr gibt wird ja bereits in
SIGNALduino_PatternExists geprüft.
Also nehme ich an, dass diese und die weiteren Regex anfragen dazu dienen den String rawdata zu zerteilen.

Wenn man die Start Sequenz nicht im Match haben will, dann packt man es einfach nicht in die capture group oder packt es in eine eigene.

Der Regex Match gibt Werte im Listen Kontext zurück 

In etwa so wäre das dann:

($start,$msgpart) = $rawmsg =~ /($startstr)($signal_str)/

Wenn man den Code aus 2316 und 2442 mit dazu nimmt könnte man die ganze suchen, teilen Aktion vereinfachen.

Dazu packen wir die Regex einfach in eine While Schleife und haben immer die gleiche Suche.
Wenn wir Mindestens und Maximallänge gleich noch mit in die Regex verpacken haben wir diese Prüfung auch gleich im Sack .

Sorry, im Handy klappt das nicht so gut. :(

Reply:
> Der Vorteil das Ergebnis aus einer Regex mit einer capture Group liegt einfach darin, das Ergebnis zu verwenden.

Bei 
`if ($rawData =~ m/$startStr/) {`
gibt es kein Ergebnis, dies ist nur eine Abfrage.

> ($start,$msgpart) = $rawmsg =~ /($startstr)($signal_str)/

Dies gibt dann aber eine größere Änderung in der SIGNALduino_Parse_MU(), dagegen sind meine bisherigen Änderungen recht klein.

http://tpcg.io/nOrsAD 
```
my $rawmsg = "67891020303020204567102030";
my $startstr = "10";
my $signal_str = "(20|30){3,}";
($start,$msgpart,$tmp,$rawmsg) = $rawmsg =~ /($startstr)($signal_str)(.*)/;
```
Mir ist aber nicht klar warum es hier die capture group $tmp gibt

Reply:
Du fragst dich, wieso Du für eine verschachtelte capture group getrennte Variablen bekommst?
Ich habe das nicht recherchiert, aber ich meine mich zu erinnern, dass Du für jede geschachtelte capture group ein Ergebnis erhälst.

Was das zurück speichern in $rawdata angeht, würde ich das vermutlich nicht machen. Und das Ergebnis in eine Variable speichern oder mit dem /g modifier arbeiten.
Ergebnisse, welche nicht benötigt werden brauchen wir nicht in eine capture group packen. Dann gibt es dafür auch keine Rückgabe .

Ich habe auch so eine Idee, dass man $rawdata komplett durch eine While Schleife in die einzelnen Wiederholungen zerteilen könnte, egal ob start definiert ist oder auch nicht.
Wenn es atart nicht gibt, müsste die Regex im 1. Durchlauf .* sein und danach auf die Sequenzen für one, Zero und Float angepasst werden.
Da bin ich aber nicht sicher, ob man eine wiederholende (/g modifier ) regex verändern kann.
Die Spezial Variablen die beschreibbar sind wir z.B. pos() lassen sich anpassen.

So langsam verstehe ich jetzt, was hier angepasst werden musste und warum. :)

Reply:
Ich habe die Änderungen commitet.
Es gibt momentan noch recht viele verbose 5 Log Ausgaben.
Ich habe 3 Ideen um diese zu reduzieren.
a) bei einigen Log Ausgaben verbose 6 verwenden

b) my $dummy = AttrVal($iohash->{NAME},"dummy",0);
und dann bei den logs `if (dummy)` anhängen.
Damit werden dann nur beim dummy sduino die vollständigen logs ausgegeben.

c) einen weiteren debuglevel 2 einführen. Bei einem Großteil der seitherigen debugausgaben wird
`if (debug == 2)` angehängt
Bei einem Großteild der log Ausgaben der Parse MU wird dann `if (debug == )` angehängt.

Reply:
Ich habe das mit verbose 6 mal getestet. Das geht nicht, daß max verbose ist 5.
Auch mit verbose:0,1,2,3,4,5,6 ist das max verbose 5

Spricht was dagegen am Anfang ein `my $dummy = AttrVal($iohash->{NAME},"dummy",0);`
und dann bei den `Log3 $name, 5, ... ` die ich zusätzlich zugefügt habe, ein` if (dummy)` anhängen.

Ein `if (debug) ` finde ich nicht so gut, da durch die debugs in SIGNALduino_PatternExists() die debug Ausgaben sehr umfangreich werden.

BTW, zu was wird die `sub SIGNALduino_MatchSignalPattern(` benötigt

Reply:
Nachtrag:
Denkbar wäre auch `$hash->{AttrList} debug:0,L,D`
L wäre dann eine erweiterte Log Ausgabe und
D wäre dann die debug Ausgabe was aus das L beinhaltet

Reply:
@sidey79 
hast Du Dir dies mal angeschaut?